tests/test_ratelimit.py

```diff
@@ -0,0 +1,137 @@
+# coding=utf-8
+# Python 2 source containing unicode https://www.python.org/dev/peps/pep-0263/
+"""
+Tests for SMTP server rate limit feature.
+
+Andrew DeOrio <awdeorio@umich.edu>
+"""
+import textwrap
+import datetime
+import time
+import future.backports.email as email
+import future.backports.email.parser  # pylint: disable=unused-import
+import pytest
+import click
+import click.testing
+from mailmerge import SendmailClient, MailmergeRateLimitError
+from mailmerge.__main__ import main
+
+try:
+    from unittest import mock  # Python 3
+except ImportError:
+    import mock  # Python 2
+
+# Python 2 pathlib support requires backport
+try:
+    from pathlib2 import Path
+except ImportError:
+    from pathlib import Path
+
+# The sh library triggers lot of false no-member errors
+# pylint: disable=no-member
+
+# We're going to use mock_SMTP because it mimics the real SMTP library
+# pylint: disable=invalid-name
+
+
+@mock.patch('smtplib.SMTP')
+def test_sendmail_ratelimit(mock_SMTP, tmp_path):
+    """Verify SMTP library calls."""
+    config_path = tmp_path/"server.conf"
+    config_path.write_text(textwrap.dedent(u"""\
+        [smtp_server]
+        host = open-smtp.example.com
+        port = 25
+        ratelimit = 60
+    """))
+    sendmail_client = SendmailClient(
+        config_path,
+        dry_run=False,
+    )
+    message = email.message_from_string(u"""
+        TO: to@test.com
+        SUBJECT: Testing mailmerge
+        FROM: from@test.com
+
+        Hello world
+    """)
+
+    # First message
+    sendmail_client.sendmail(
+        sender="from@test.com",
+        recipients=["to@test.com"],
+        message=message,
+    )
+    smtp = mock_SMTP.return_value.__enter__.return_value
+    assert smtp.sendmail.call_count == 1
+
+    # Second message exceeds the rate limit, doesn't try to send a message
+    with pytest.raises(MailmergeRateLimitError):
+        sendmail_client.sendmail(
+            sender="from@test.com",
+            recipients=["to@test.com"],
+            message=message,
+        )
+    assert smtp.sendmail.call_count == 1
+
+    # Retry the second message after 1 s because the rate limit is 60 messages
+    # per minute
+    # FIXME a better way to do this is to mock datetime.datetime.now()
```

Comment:
I'm hesitant to leave a `time.sleep()` in a test.  Any suggestions?  Ever used [time_machine](https://pypi.org/project/time-machine/) ?

Reply:
Fixed! 1e54a542dcb5638aaa8bb4f90c1e571e599a8207